homeassistant/components/recorder/__init__.py

```diff
@@ -191,38 +199,25 @@ def run(self):
                 self.queue.task_done()
                 return
 
-            elif event.event_type == EVENT_TIME_CHANGED:
+            if event.event_type == EVENT_TIME_CHANGED:
                 self.queue.task_done()
                 continue
 
             session = Session()
             dbevent = Events.from_event(event)
             session.add(dbevent)
-
-            for _ in range(0, RETRIES):
-                try:
-                    session.commit()
-                    break
-                except sqlalchemy.exc.OperationalError as e:
-                    log_error(e, retry_wait=QUERY_RETRY_WAIT,
-                              rollback=True)
+            self._commit(session)
 
             if event.event_type != EVENT_STATE_CHANGED:
                 self.queue.task_done()
                 continue
 
+            # Can possibly be combined into single session with the above
             session = Session()
             dbstate = States.from_event(event)
-
-            for _ in range(0, RETRIES):
-                try:
-                    dbstate.event_id = dbevent.event_id
-                    session.add(dbstate)
-                    session.commit()
-                    break
-                except sqlalchemy.exc.OperationalError as e:
-                    log_error(e, retry_wait=QUERY_RETRY_WAIT,
-                              rollback=True)
+            dbstate.event_id = dbevent.event_id
+            session.add(dbstate)
```

Comment:
I think (2) is cleaner (although the function name might need to be better) as it solves the repeated code problem you already addressed.  To test whether retries work, I used this approach:

- start HA
- let it record some events
- tap on an item in the UI to view it's state history
- stop and start mysql
- repeat the above to ensure that events keep being stored and that tapping on the UI works without failure